feat: import ovf template from url in supervisor builder

ericvmw opened the pull request:
### Summary

A new feature is being added to Supervisors to expose a new custom resource [`ContentLibraryItemImportRequest`](https://github.com/vmware-tanzu/image-registry-operator-api/blob/main/api/v1alpha1/contentlibraryitemimportrequest_types.go) that allows users to import OVF templates to a namespace with writable & import-allowed content library configured on a Supervisor, this enables users to use Packer to orchestrate a complete e2e workflow, from importing their source image from HTTP server, to customize and publish the image to Supervisor clusters.

This change adds configurable optional steps on supervisor builder plugin to allow users to import OVF template to Supervisor namespaces via packer builder vsphere supervisor plugin.

1. Add import image step to import image from remote URL to the target content library in the specified namespace on the Supervisor by creating ContentLibraryItemImportRequest resource, wait until the import request completes. If keep_input_artifact is set to be true, the import request will be cleaned after it is done. Besides taking the **image_name** config from existing **CreateSource** step, this step also support the following optional configs: 
**import_source_url**: the remote URL to import OVF template from, must starts with https. 
**import_source_ssl_certificate**: the SSL certificate of the remote HTTPS server. 
**import_target_location_name**: the import target location (content library that must be writable and allow import).
**import_target_image_type**: optional, the target image type, it defaults to OVF type and only OVF type is supported currently.
**import_request_name**: The name of the ContentLibraryItemImportRequest resource, otherwise it defaults to `packer-vsphere-supervisor-import-req-<random-suffix>`. 
**watch_import_timeout_sec**: The timeout in seconds to wait for the image to be imported, otherwise it defaults to `600`.
**clean_imported_image**: optional, whether to clean the image imported in this step. If it is set to true, the imported image will be deleted after source VM is created and becomes ready. Defaults to false.

2. Add unit tests to cover both import validation and import image steps.

Note that in order to import OVF template from remote URL, the target content library must be configured with the default OVF security policy, and the OVF template must be signed with a certificated that is signed by a CA cert trusted by Content Library Service on vCenter, otherwise the import OVF template image will not be security compliant and cannot be deployed.

### Testing

```sh
Testing Done:
~/Git/github.com/hashicorp/packer-plugin-vsphere   feat/import-ovf-from-url-on-supervisor
 $ go fmt ./...

~/Git/github.com/hashicorp/packer-plugin-vsphere   feat/import-ovf-from-url-on-supervisor
 $ make generate
2024/05/15 10:36:18 Copying "docs" to ".docs/"
2024/05/15 10:36:18 Replacing @include '...' calls in .docs/ Compiling MDX docs in '.docs' to Markdown in '.web-docs'...

~/Git/github.com/hashicorp/packer-plugin-vsphere   feat/import-ovf-from-url-on-supervisor
 $ make build

~/Git/github.com/hashicorp/packer-plugin-vsphere   feat/import-ovf-from-url-on-supervisor
 $ make test
?   	github.com/hashicorp/packer-plugin-vsphere	[no test files]
?   	github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/common/testing	[no test files]
?   	github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/examples/driver	[no test files]
?   	github.com/hashicorp/packer-plugin-vsphere/version	[no test files]
ok  	github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/clone	2.033s
ok  	github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/common	3.404s
ok  	github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/driver	4.862s
ok  	github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/iso	2.362s
ok  	github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/supervisor	5.488s
ok  	github.com/hashicorp/packer-plugin-vsphere/post-processor/vsphere	1.651s
ok  	github.com/hashicorp/packer-plugin-vsphere/post-processor/vsphere-template	2.766s
```

added unit tests pass:

```
=== RUN   TestImportImage_Prepare
--- PASS: TestImportImage_Prepare (0.00s)
=== RUN   TestStepImportImage_Run_Skip
2024/05/17 14:58:41 ui: Skipping image import step. Required configurations for the import are not set.
2024/05/17 14:58:41 ui: Skipping image import step. Required configurations for the import are not set.
--- PASS: TestStepImportImage_Run_Skip (0.00s)
=== RUN   TestStepImportImage_Run_Validate
2024/05/17 14:58:41 ui error: error checking required states: missing required state: supervisor_namespace
2024/05/17 14:58:41 ui error: failed to initialize image import: missing required state: supervisor_namespace
2024/05/17 14:58:41 ui error: failed to initialize image import: failed to cast kube_client to type client.WithWatch
2024/05/17 14:58:41 ui: Validating image import request...
2024/05/17 14:58:41 ui error: failed to validate import image configs: import request source url certificate is empty
2024/05/17 14:58:41 ui: Validating image import request...
2024/05/17 14:58:41 ui error: failed to return the content library by name cl-6066c61f7931c5ef9 in namespace test-ns
2024/05/17 14:58:41 ui error: failed to validate import image configs: contentlibraries.imageregistry.vmware.com "cl-6066c61f7931c5ef9" not found
2024/05/17 14:58:41 ui: Validating image import request...
2024/05/17 14:58:41 ui error: failed to validate import image configs: import target content library "cl-6066c61f7931c5ef9" is not writable or does not allow import
2024/05/17 14:58:41 ui: Validating image import request...
2024/05/17 14:58:41 ui error: failed to validate import image configs: image type other is not supported
--- PASS: TestStepImportImage_Run_Validate (0.00s)
=== RUN   TestStepImportImage_Run
2024/05/17 14:58:41 ui: Validating image import request...
2024/05/17 14:58:41 ui: Image import request source and target are valid.
2024/05/17 14:58:41 ui: Importing the source image from https://example.com/example.ovf to cl-6066c61f7931c5ef9.
2024/05/17 14:58:41 ui: Creating ContentLibraryItemImportRequest object test-req-name in namespace test-ns.
2024/05/17 14:58:41 ui: Successfully created the ContentLibraryItemImportRequest object test-req-name.
2024/05/17 14:58:42 ui: Successfully imported the image as a content library item &{"imageregistry.vmware.com/v1alpha1" "ContentLibraryItem" "clitem-xxx"}.
2024/05/17 14:58:42 ui: Finished importing the image from https://example.com/example.ovf to cl-6066c61f7931c5ef9.
--- PASS: TestStepImportImage_Run (1.00s)
=== RUN   TestStepImportImage_Cleanup
2024/05/17 14:58:42 ui: Skipping clean up of the ContentLibraryItemImportRequest object as specified in config.
2024/05/17 14:58:42 ui: Deleting the ContentLibraryItemImportRequest object test-req-name from Supervisor cluster.
2024/05/17 14:58:42 ui: Successfully deleted the ContentLibraryItemImportRequest object.
--- PASS: TestStepImportImage_Cleanup (0.00s)

$ make testacc
all acceptance tests pass except some tests under github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/iso, which doesn't look related to the change.

Supervisor packer build pass with import_source URL: https://wp-content-pstg.broadcom.com/vmimagesvc/ubuntu-jammy-22.04-cloudimg-signed/ubuntu-jammy-22.04-cloudimg-signed.ovf

packer build --var-file tmp/packer-example.json tmp/packer-example.pkr.hcl vsphere-supervisor.vm: 

==> vsphere-supervisor.vm: Creating temporary RSA SSH key for instance...
    vsphere-supervisor.vm: Connecting to Supervisor cluster...
    vsphere-supervisor.vm: Successfully connected to Supervisor cluster
    vsphere-supervisor.vm: Validating VM publish location...
    vsphere-supervisor.vm: VM publish location is valid
    vsphere-supervisor.vm: Validating image import request...
    vsphere-supervisor.vm: Image import request source and target are valid
    vsphere-supervisor.vm: Importing the source image from https://wp-content-pstg.broadcom.com/vmimagesvc/ubuntu-jammy-22.04-cloudimg-signed/ubuntu-jammy-22.04-cloudimg-signed.ovf to cl-6066c61f7931c5ef9
    vsphere-supervisor.vm: Creating a ContentLibraryItemImportRequest object <sensitive>-vsphere-supervisor-import-req-71b5a
    vsphere-supervisor.vm: Successfully created the ContentLibraryItemImportRequest object <sensitive>-vsphere-supervisor-import-req-71b5a
    vsphere-supervisor.vm: Waiting for the image import request to complete...
    vsphere-supervisor.vm: Waiting for the image import request to complete...
    vsphere-supervisor.vm: Successfully imported the image as library item &{"imageregistry.<sensitive>.com/v1alpha1" "ContentLibraryItem" "clitem-c950e9d6dc3b57ee4"}
    vsphere-supervisor.vm: Finished importing the image from https://wp-content-pstg.broadcom.com/vmimagesvc/ubuntu-jammy-22.04-cloudimg-signed/ubuntu-jammy-22.04-cloudimg-signed.ovf to "cl-6066c61f7931c5ef9"
    vsphere-supervisor.vm: Creating required source objects in Supervisor cluster...
    vsphere-supervisor.vm: Creating a K8s Secret object for providing source VM bootstrap data...
    vsphere-supervisor.vm: Using default cloud-init user data as the 'bootstrap_data_file' is not specified
    vsphere-supervisor.vm: Successfully created the K8s Secret object
    vsphere-supervisor.vm: Creating a source VirtualMachine object
    vsphere-supervisor.vm: Successfully created the VirtualMachine object
    vsphere-supervisor.vm: Creating a VirtualMachineService object for network connection
    vsphere-supervisor.vm: Successfully created the VirtualMachineService object
    vsphere-supervisor.vm: Finished creating all required source objects in Supervisor cluster
    vsphere-supervisor.vm: Waiting for the source VM to be powered-on and accessible...
    vsphere-supervisor.vm: Source VM is NOT powered-on yet, continue watching...
    vsphere-supervisor.vm: Source VM is NOT powered-on yet, continue watching...
    vsphere-supervisor.vm: Source VM is powered-on, waiting for an IP to be assigned...
    vsphere-supervisor.vm: Source VM is powered-on, waiting for an IP to be assigned...
    vsphere-supervisor.vm: Successfully obtained the source VM IP: 192.168.128.189
    vsphere-supervisor.vm: Getting source VM ingress IP from the VMService object
    vsphere-supervisor.vm: Successfully retrieved the source VM ingress IP: 192.168.0.4
    vsphere-supervisor.vm: Source VM is now ready in Supervisor cluster
==> vsphere-supervisor.vm: Using SSH communicator to connect: 192.168.0.4 ==> vsphere-supervisor.vm: Waiting for SSH to become available... ==> vsphere-supervisor.vm: Connected to SSH!
==> vsphere-supervisor.vm: Provisioning with shell script: /var/folders/gc/ptc0zb0507b05r97rqnb5cym0000gq/T/<sensitive>-shell1862650776
    vsphere-supervisor.vm: Publishing the source VM to "cl-6066c61f7931c5ef9"
    vsphere-supervisor.vm: Creating a VirtualMachinePublishRequest object
    vsphere-supervisor.vm: Successfully created the VirtualMachinePublishRequest object
    vsphere-supervisor.vm: Waiting for the VM publish request to complete...
    vsphere-supervisor.vm: Waiting for the VM publish request to complete...
    vsphere-supervisor.vm: Successfully published the VM to image "vmi-2569621448a150e8a"
    vsphere-supervisor.vm: Finished publishing the source VM
    vsphere-supervisor.vm: Deleting the VirtualMachinePublishRequest object from Supervisor cluster
    vsphere-supervisor.vm: Successfully deleted the VirtualMachinePublishRequest object
    vsphere-supervisor.vm: Deleting the imported ContentLibraryItem object clitem-19fab321189b7b177 in namespace test-k8s-service.
    vsphere-supervisor.vm: Successfully deleted the ContentLibraryItem object clitem-19fab321189b7b177 in namespace test-k8s-service.
    vsphere-supervisor.vm: Deleting the VirtualMachineService object from Supervisor cluster
    vsphere-supervisor.vm: Successfully deleted the VirtualMachineService object
    vsphere-supervisor.vm: Deleting the VirtualMachine object from Supervisor cluster
    vsphere-supervisor.vm: Successfully deleted the VirtualMachine object
    vsphere-supervisor.vm: Deleting the K8s Secret object from Supervisor cluster
    vsphere-supervisor.vm: Successfully deleted the K8s Secret object
    vsphere-supervisor.vm: Deleting the ContentLibraryItemImportRequest object <sensitive>-vsphere-supervisor-import-req-71b5a from Supervisor cluster
    vsphere-supervisor.vm: Successfully deleted the ContentLibraryItemImportRequest object
    vsphere-supervisor.vm: Build 'vsphere-supervisor' finished successfully.
Build 'vsphere-supervisor.vm' finished after 5 minutes 3 seconds.

Supervisor packer build pass without setting import_source_url (use existing image instead)
```

### Reference

Closes #434

tenthirtyam reviewed (changes requested):
Thanks, Eric!

Per the README.md, in case of feature contribution, we kindly ask you to open an issue to discuss it beforehand so that these can be linked.

Also, please update `https://github.com/hashicorp/packer-plugin-vsphere/blob/main/docs/builders/vsphere-supervisor.mdx` to include the new items as this will then populate webdocs when `make generate` is run.

ericvmw commented:
> Thanks, Eric!
> 
> * Per the README.md, in case of feature contribution, we kindly ask you to open an issue to discuss it beforehand so that these can be linked.
> * Could you use the Conventional Commits style and ensure it is signed off with your broadcom.com email attached to your GitHub account. The commit message should be short and you can leave this content in the linked PR.
> * Also, please update `https://github.com/hashicorp/packer-plugin-vsphere/blob/main/docs/builders/vsphere-supervisor.mdx` to include the new items as this will then populate webdocs when `make generate` is run.

Thanks Ryan! Updated as below:
1. Opened the issue here: https://github.com/hashicorp/packer-plugin-vsphere/issues/434.
2. Shortened the commit message and included my broadcom.com email.
3. Updated the vsphere-supervisor.mdx, regenerated the webdocs, and fixed lint issues.

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+type ImportImageConfig struct {
+	// The name of the image import request. If not specified, the name defaults to `packer-vsphere-supervisor-import-req-<random-suffix>`.
+	ImportRequestName string `mapstructure:"import_request_name"`
+	// The timeout in seconds to wait for the image to be imported. If not specified, it defaults to `600`.
```
```suggestion
	// The timeout in seconds to wait for the image to be imported. 
	// Defaults to `600`.
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+var IsWatchingImageImport bool
+
+type ImportImageConfig struct {
+	// The name of the image import request. If not specified, the name defaults to `packer-vsphere-supervisor-import-req-<random-suffix>`.
```
```suggestion
	// The name of the image import request.
	// Defaults to `packer-vsphere-supervisor-import-req-<random-suffix>`.
```

ericvmw replied:
Thanks, will update to this format for all configs.

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+		c.WatchImportTimeoutSec = DefaultWatchImageImportTimeoutSec
+	}
+	if c.ImportRequestName == "" {
+		c.ImportRequestName = "packer-vsphere-supervisor-import-req-" + uuid.NewString()[:5]
```
Could use a const here.

ericvmw replied:
Yep, declared a const instant

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+func (s *StepImportImage) Cleanup(state multistep.StateBag) {
+	if state.Get(StateKeyImageImportRequestCreated) == false {
+		// Either the image import step was skipped or the object was not created successfully.
+		// Skip deleting the Content Library import Request object.
```
```suggestion
		// Skip deleting the content library import request object.
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+		},
+	}
+	if err := s.KubeWatchClient.Delete(ctx, itemImportReqObj); err != nil {
+		logger.Error("Failed to delete the ContentLibraryItemImportRequest object: %s", err)
```
```suggestion
		logger.Error("error deleting the ContentLibraryItemImportRequest object: %s", err)
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+	if err := s.KubeWatchClient.Delete(ctx, itemImportReqObj); err != nil {
+		logger.Error("Failed to delete the ContentLibraryItemImportRequest object: %s", err)
+	} else {
+		logger.Info("Successfully deleted the ContentLibraryItemImportRequest object")
```
```suggestion
		logger.Info("Successfully deleted the ContentLibraryItemImportRequest object.")
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+		StateKeySupervisorNamespace,
+		StateKeyKubeClient,
+	); err != nil {
+		logger.Error("failed check required states: %s", err)
```
```suggestion
		logger.Error("error checking required states: %s", err)
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+	"github.com/google/uuid"
+
+	"github.com/hashicorp/packer-plugin-sdk/multistep"
+	"github.com/vmware-tanzu/image-registry-operator-api/api/v1alpha1"
```
like `imagev1` or `imageRegistryv1`.

ericvmw replied:
yep, makes sense, update to imgregv1 in all files.

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+
+	// Skip importing if the import source URL or target location is not specified.
+	if s.SourceURL == "" || s.TargetLocationName == "" {
+		logger.Info("Skip importing image since the source URL or target location is not specified.")
```
```suggestion
		logger.Info("Skipping image import. Source URL or target location is not specified.")
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+		return multistep.ActionHalt
+	}
+
+	logger.Info("Finished importing the image from %s to %q", s.SourceURL, s.TargetLocationName)
```
```suggestion
	logger.Info("Finished importing the image from %s to %q.", s.SourceURL, s.TargetLocationName)
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+		return
+	}
+
+	logger.Info(fmt.Sprintf("Deleting the ContentLibraryItemImportRequest object %s from Supervisor cluster", s.ImportImageConfig.ImportRequestName))
```
```suggestion
	logger.Info(fmt.Sprintf("Deleting the ContentLibraryItemImportRequest object %s from Supervisor cluster.", s.ImportImageConfig.ImportRequestName))
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+func (s *StepImportImage) createImageImportRequest(ctx context.Context, logger *PackerLogger) error {
+	logger.Info("Creating a ContentLibraryItemImportRequest object %s", s.ImportImageConfig.ImportRequestName)
```
```suggestion
	logger.Info("Creating ContentLibraryItemImportRequest object %s.", s.ImportImageConfig.ImportRequestName)
```

ericvmw replied:
Done

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	if err := s.KubeWatchClient.Create(ctx, imageImportReq); err != nil {
+		logger.Error("Failed to create the ContentLibraryItemImportRequest object %s", imageImportReq.Name)
```
```suggestion
		logger.Error("error creating the ContentLibraryItemImportRequest object %s", imageImportReq.Name)
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+		return err
+	}
+
+	logger.Info("Successfully created the ContentLibraryItemImportRequest object %s", imageImportReq.Name)
```
```suggestion
	logger.Info("Successfully created the ContentLibraryItemImportRequest object %s.", imageImportReq.Name)
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+	})
+
+	if err != nil {
+		logger.Error("Failed to watch the ContentLibraryItemImportRequest object in Supervisor cluster")
```
```suggestion
		logger.Error("error watching the ContentLibraryItemImportRequest object in supervisor cluster")
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+				}
+			}
+			if importSuccess {
+				logger.Info("Successfully imported the image as library item %q", itemImportReqObj.Status.ItemRef)
```
```suggestion
				logger.Info("Successfully imported the image as a content library item %q.", itemImportReqObj.Status.ItemRef)
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image_test.go:
```diff
@@ -0,0 +1,222 @@
+	"time"
+
+	"github.com/hashicorp/packer-plugin-sdk/multistep"
+	"github.com/vmware-tanzu/image-registry-operator-api/api/v1alpha1"
```
I think it would be better to alias this so that les changes are needed in the future when this is past alpha.

ericvmw replied:
Yep, updated to imgregv1

tenthirtyam commented on builder/vsphere/supervisor/step_import_image_test.go:
```diff
@@ -0,0 +1,222 @@
+	action := step.Run(context.TODO(), state)
+	if action != multistep.ActionContinue {
+		if rawErr, ok := state.GetOk("error"); ok {
+			t.Errorf("Error from running the step: %s", rawErr.(error))
```
```suggestion
			t.Errorf("error running step: %s", rawErr.(error))
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image_test.go:
```diff
@@ -0,0 +1,222 @@
+		action := step.Run(ctx, state)
+		if action == multistep.ActionHalt {
+			if rawErr, ok := state.GetOk("error"); ok {
+				t.Errorf("Error from running the step: %s", rawErr.(error))
```
```suggestion
				t.Errorf("error running step: %s", rawErr.(error))
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image_test.go:
```diff
@@ -0,0 +1,222 @@
+			if rawErr, ok := state.GetOk("error"); ok {
+				t.Errorf("Error from running the step: %s", rawErr.(error))
+			}
+			t.Errorf("Step should NOT halt")
```
```suggestion
			t.Errorf("step should not halt")
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image_test.go:
```diff
@@ -0,0 +1,222 @@
+		}
+		importReq := &v1alpha1.ContentLibraryItemImportRequest{}
+		if err := testKubeClient.Get(ctx, objKey, importReq); err != nil {
+			t.Errorf("Failed to get the expected ContentLibraryItemImportRequest object, err: %s", err.Error())
```
```suggestion
			t.Errorf("failed to return the expected ContentLibraryItemImportRequest object, err: %s", err.Error())
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image_test.go:
```diff
@@ -0,0 +1,222 @@
+	}
+	importReq := &v1alpha1.ContentLibraryItemImportRequest{}
+	if err := testKubeClient.Get(ctx, objKey, importReq); err != nil {
+		t.Errorf("Failed to get the expected ContentLibraryItemImportRequest object, err: %s", err.Error())
```
```suggestion
		t.Errorf("railed to return the expected ContentLibraryItemImportRequest object, err: %s", err.Error())
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image_test.go:
```diff
@@ -0,0 +1,222 @@
+		Name:       testCLItemName,
+	}
+	if err := testKubeClient.Update(ctx, importReq); err != nil {
+		t.Errorf("Failed to update the ContentLibraryItemImportRequest object status image name, err: %s", err.Error())
```
```suggestion
		t.Errorf("failed to update the ContentLibraryItemImportRequest object status image name, err: %s", err.Error())
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image_test.go:
```diff
@@ -0,0 +1,222 @@
+	state.Put(supervisor.StateKeyImageImportRequestCreated, true)
+	step.Cleanup(state)
+
+	expectedOutput := []string{"Skip cleaning up the ContentLibraryItemImportRequest object as specified in config"}
```
```suggestion
	expectedOutput := []string{"Skipping clean up of the ContentLibraryItemImportRequest object as specified in config."}
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,174 @@
+	"strings"
+
+	"github.com/hashicorp/packer-plugin-sdk/multistep"
+	imgregv1a1 "github.com/vmware-tanzu/image-registry-operator-api/api/v1alpha1"
```
I think it would be better to have a common alias for this.

ericvmw replied:
changed to imgregv1

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,174 @@
+	// The remote URL where the to-be-imported image is hosted, the URL must start with `https`.
+	ImportSourceURL string `mapstructure:"import_source_url"`
+	// The SSL certificate of the remote HTTPs server that hosts the to-be-imported image.
+	ImportSourceSSLCertificate string `mapstructure:"import_source_ssl_certificate"`
+
+	// Name of a writable & import-allowed ContentLibrary resource in the namespace where the image will be imported to.
+	ImportTargetLocationName string `mapstructure:"import_target_location_name"`
```
```suggestion
	ImportSourceSSLCertificate string `mapstructure:"import_source_ssl_certificate"`
	// Name of a writable and import-allowed content library resource in the namespace where the image will be imported.
	ImportTargetLocationName string `mapstructure:"import_target_location_name"`
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,174 @@
+	state.Put(StateKeyImportTargetImageType, s.Config.ImportTargetImageType)
+
+	if s.Config.ImportSourceURL == "" || s.Config.ImportTargetLocationName == "" {
+		logger.Info("image import step will be skipped as the required configs for the import is not set")
```
```suggestion
		logger.Info("Skipping image import step. Required configurations for the import are not set.")
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,174 @@
+	// Use dry run mode to send an image import creation request to API-Server without applying the resource.
+	err := client.NewDryRunClient(s.KubeClient).Create(ctx, importReq)
+	if err != nil && strings.Contains(err.Error(), importOVFFeatureNotEnabledMsg) {
+		logger.Error("The image import feature is NOT enabled in the current version of vSphere Supervisor cluster")
```
```suggestion
		logger.Error("image import feature is not enabled")
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,174 @@
+
+func (s *StepValidateImport) isImportSourceValid() error {
+	if s.Config.ImportSourceURL == "" || !strings.HasPrefix(s.Config.ImportSourceURL, "https://") {
+		return fmt.Errorf("the specified import request source URL is either empty or does not start with https")
```
```suggestion
		return fmt.Errorf("the specified import request source url is either empty or does not start with https")
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,174 @@
+		return fmt.Errorf("the specified import request source URL is either empty or does not start with https")
+	}
+	if s.Config.ImportSourceSSLCertificate == "" {
+		return fmt.Errorf("the specified import request source URL certificate is empty")
```
```suggestion
		return fmt.Errorf("import request source url certificate is empty")
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,174 @@
+	cl := &imgregv1a1.ContentLibrary{}
+	objKey := client.ObjectKey{Name: s.Config.ImportTargetLocationName, Namespace: s.Namespace}
+	if err := s.KubeClient.Get(ctx, objKey, cl); err != nil {
+		logger.Error(fmt.Sprintf("failed to get ContentLibrary by name %s in namespace %s", s.Config.ImportTargetLocationName, s.Namespace))
```
```suggestion
		logger.Error(fmt.Sprintf("failed to return content library by name %s in namespace %s", s.Config.ImportTargetLocationName, s.Namespace))
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	if !cl.Spec.Writable || !cl.Spec.AllowImport {
+		return fmt.Errorf("the specified import target content library %q is not writable or does not allow import", s.Config.ImportTargetLocationName)
```
```suggestion
		return fmt.Errorf("import target content library %q is not writable or does not allow import", s.Config.ImportTargetLocationName)
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,174 @@
+
+func (c *ValidateImportConfig) Prepare() []error {
+	if c.ImportTargetImageType == "" {
+		c.ImportTargetImageType = "OVF"
```
```suggestion
		c.ImportTargetImageType = "ovf"
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,174 @@
+	ImportTargetImageType string `mapstructure:"import_target_image_type"`
+}
+
+func (c *ValidateImportConfig) Prepare() []error {
+	if c.ImportTargetImageType == "" {
+		c.ImportTargetImageType = "OVF"
+	}
+
+	return nil
+}
```
```suggestion
func (c *ValidateImportConfig) Prepare() []error {
    switch c.ImportTargetImageType {
    case "":
        c.ImportTargetImageType = "ovf"
    case "ovf":
        // If it's already "ovf", do nothing.
    default:
        return []error{fmt.Errorf("Unsupported ImportTargetImageType: %s", c.ImportTargetImageType)}
    }

    return nil
}
```

ericvmw replied:
Thanks, it is better, updated!

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+	action := step.Run(ctx, state)
+	if action != multistep.ActionContinue {
+		if rawErr, ok := state.GetOk("error"); ok {
+			t.Errorf("Error from running the step: %s", rawErr.(error))
```
```suggestion
			t.Errorf("error running step: %s", rawErr.(error))
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+	}
+
+	if config.ImportTargetImageType != "OVF" {
+		t.Fatal("The default import target image type should be OVF")
```
```suggestion
		t.Fatal("The default import target image type should be 'ovf'.")
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+		t.Fatalf("Prepare should not fail: %v", actualErr)
+	}
+
+	if config.ImportTargetImageType != "OVF" {
```
```suggestion
	if config.ImportTargetImageType != "ovf" {
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+	"testing"
+
+	"github.com/hashicorp/packer-plugin-sdk/multistep"
+	imgregv1a1 "github.com/vmware-tanzu/image-registry-operator-api/api/v1alpha1"
```
This should be consistent. WOuld not recommend the a1.

`imagev1` or `imageRegistryv1`

ericvmw replied:
changed to imgregv1

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+	action = step.Run(ctx, state)
+	if action != multistep.ActionContinue {
+		if rawErr, ok := state.GetOk("error"); ok {
+			t.Errorf("Error from running the step: %s", rawErr.(error))
```
```suggestion
			t.Errorf("error running step: %s", rawErr.(error))
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+
+	if rawErr, ok := state.GetOk("error"); ok {
+		if rawErr.(error).Error() != "missing required state: supervisor_namespace" {
+			t.Errorf("Error from running the step doesn't match: %s", rawErr.(error))
```
```suggestion
			t.Errorf("error running step: %s", rawErr.(error))
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+	expectedError := "failed to cast \"kube_client\" from state bag as type 'client.Client'"
+	if rawErr, ok := state.GetOk("error"); ok {
+		if !strings.Contains(rawErr.(error).Error(), expectedError) {
+			t.Errorf("Expected error contains %v, but got %v", expectedError, rawErr.(error).Error())
```
```suggestion
			t.Errorf("expected error contains %v, but got %v", expectedError, rawErr.(error).Error())
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+	checkOutputLines(t, testWriter, expectedOutput)
+
+	// 3. Test with unsupported source URL.
+	step.Config.ImportSourceURL = "http://test.com/test.ovf"
```
```suggestion
	step.Config.ImportSourceURL = "http://example.com/example.ovf"
```

ericvmw replied:
updated it.

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+		t.Fatal("Step should halt")
+	}
+
+	expectedError = "the specified import request source URL is either empty or does not start with https"
```
```suggestion
	expectedError = "import request source url is either empty or does not start with https"
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+	expectedError = "the specified import request source URL is either empty or does not start with https"
+	if rawErr, ok := state.GetOk("error"); ok {
+		if !strings.Contains(rawErr.(error).Error(), expectedError) {
+			t.Errorf("Expected error contains %v, but got %v", expectedError, rawErr.(error).Error())
```
```suggestion
			t.Errorf("expected error contains %v, but got %v", expectedError, rawErr.(error).Error())
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+		t.Fatal("Step should halt")
+	}
+
+	expectedError = "the specified import request source URL certificate is empty"
```
```suggestion
	expectedError = "import request source url certificate is empty"
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+	expectedError = "the specified import request source URL certificate is empty"
+	if rawErr, ok := state.GetOk("error"); ok {
+		if !strings.Contains(rawErr.(error).Error(), expectedError) {
+			t.Errorf("Expected error contains %v, but got %v", expectedError, rawErr.(error).Error())
```
```suggestion
			t.Errorf("expected error contains %v, but got %v", expectedError, rawErr.(error).Error())
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+	expectedError = fmt.Sprintf("contentlibraries.imageregistry.vmware.com \"%s\" not found", testTargetLibrary)
+	if rawErr, ok := state.GetOk("error"); ok {
+		if !strings.Contains(rawErr.(error).Error(), expectedError) {
+			t.Errorf("Expected error contains %v, but got %v", expectedError, rawErr.(error).Error())
```
```suggestion
			t.Errorf("expected error contains %v, but got %v", expectedError, rawErr.(error).Error())
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+
+	expectedOutput = []string{
+		"Validating image import request...",
+		fmt.Sprintf("failed to get ContentLibrary by name %s in namespace %s", testTargetLibrary, testNamespace),
```
```suggestion
		fmt.Sprintf("failed to return the content library by name %s in namespace %s", testTargetLibrary, testNamespace),
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+		t.Fatal("Step should halt")
+	}
+
+	expectedError = fmt.Sprintf("the specified import target content library %q is not writable or does not allow import", testTargetLibrary)
```
```suggestion
	expectedError = fmt.Sprintf("import target content library %q is not writable or does not allow import", testTargetLibrary)
```

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import_test.go:
```diff
@@ -0,0 +1,265 @@
+	if action != multistep.ActionHalt {
+		t.Fatal("Step should halt")
+	}
+	expectedError = "the given target image type other is not supported"
```
```suggestion
	expectedError = "image type other is not supported"
```

tenthirtyam commented on docs/builders/vsphere-supervisor.mdx:
```diff
@@ -78,6 +89,14 @@ There are various configuration options available for each step in this builder.
 
 @include 'builder/vsphere/supervisor/ConnectSupervisorConfig-not-required.mdx'
 
+#### Source VM Image Import Validation
+
+@include 'builder/vsphere/supervisor/ValidateImportConfig-not-required.mdx'
+
+#### Source VM Image Importing
+
+@include 'builder/vsphere/supervisor/ImportImageConfig-not-required.mdx'
```
```suggestion
#### Source VM Image Import Validation

**Optional:**

@include 'builder/vsphere/supervisor/ValidateImportConfig-not-required.mdx'

#### Source VM Image Importing

**Optional:**

@include 'builder/vsphere/supervisor/ImportImageConfig-not-required.mdx'
```

ericvmw replied:
Thanks, this `optional` helps!

tenthirtyam commented on go.mod:
```diff
@@ -1,13 +1,15 @@
 module github.com/hashicorp/packer-plugin-vsphere
 
-go 1.19
+go 1.21
```
🤔 `.go-version` is set at 1.21.8.

Should this be:

go 1.21

toolchain go1.12.8

❔

ericvmw replied:
Good catch, yeah, should be 1.21.8

tenthirtyam reviewed (changes requested):
🦖 Provided an initial review.

My comment mostly just nits that would come back in a `staticcheck ./...` (#435), conforming to an upcoming docs update (#430), etc.

A couple other minor suggestions otherwise.

ericvmw commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,260 @@
+		return multistep.ActionContinue
+	}
+
+	logger.Info("Importing the source image from %s to %s", s.SourceURL, s.TargetLocationName)
```
Thanks, adding closing period to all info log messages.

ericvmw reviewed (commented):
Thanks for the review!

tenthirtyam commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,179 @@
+	case "ovf":
+		// If it's already "ovf", do nothing.
+	default:
+		return []error{fmt.Errorf("Unsupported ImportTargetImageType: %s", c.ImportTargetImageType)}
```
```suggestion
		return []error{fmt.Errorf("unsupported ImportTargetImageType: %s", c.ImportTargetImageType)}
```

tenthirtyam reviewed (changes requested):
Just one more minor nit from `staticcheck` and marking all conversations as closed.

Looks good, Eric. 🎉

dilyar85 commented on builder/vsphere/supervisor/builder.go:
```diff
@@ -57,6 +57,22 @@ func (b *Builder) Run(ctx context.Context, ui packersdk.Ui, hook packersdk.Hook)
+	// conditionally add steps to validate import spec and import images from source URL as VM image.
+	if b.config.ValidateImportConfig.ImportSourceURL != "" {
+		steps = append(steps,
+			&StepValidateImport{
+				Config: &b.config.ValidateImportConfig,
+			},
```
After looking at the implementation of this step, I kind of feel like we may not need it in this plugin. Most of its functionality duplicates the CRD scheme check and webhook validation. I am concerned that if we change those in the future (which is likely when we add ISO import support, for example), we would need to update this step as well.

The reason we have the `ValidatePublish` step is that a VM is published at the end, which could lead to a poor user experience if errors in their publish configs are discovered after the VM is deployed and customized. For import, since it's executed at a pretty early stage, I think it's sufficient to perform some basic config checks in the `Prepare()` function and let the image-registry-operator controller and webhook handle the rest. The Packer plugin could simply output a message if any error occurs.

ericvmw replied:
I think this is a fair point, we don't have to maintain 2 steps, I'm combining them into one step instead, and make it easier to maintain.

dilyar85 commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,264 @@
+
+var IsWatchingImageImport bool
+
+type ImportImageConfig struct {
```
It would be nice if we could also provide an option to keep the import source image (true by default). If set to false, it would allow users to clean up the imported image after deploying a VM from it.

ericvmw replied:
Hm... it could be useful, I think it doesn't belong there though, it better be in the CreateSource step if to add it. I'll leave it as it is for now, may add the option if I get some cycle to do it.

dilyar85 commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,264 @@
+		return multistep.ActionHalt
+	}
+
+	// Skip importing if the import source URL or target location is not specified.
+	if s.SourceURL == "" || s.TargetLocationName == "" {
+		logger.Info("Skipping image import. Source URL or target location is not specified.")
+		return multistep.ActionContinue
+	}
```
Can be removed since the step is added when these conditions are false. Plus, the `s.initStep()` can also guarantee that these configs are set.

ericvmw replied:
Keeping this one by updating other places.

dilyar85 commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,264 @@
+}
+
+func (s *StepImportImage) createImageImportRequest(ctx context.Context, logger *PackerLogger) error {
+	logger.Info("Creating ContentLibraryItemImportRequest object %s.", s.ImportImageConfig.ImportRequestName)
```
Nit: Add namespace to the log as well?

ericvmw replied:
Sounds good.

tenthirtyam commented on .web-docs/components/builder/vsphere-supervisor/README.md:
```diff
@@ -83,6 +94,39 @@ There are various configuration options available for each step in this builder.
 
+#### Source VM Image Import Validation
+
+**Optional:**
```
This layout will be in an upcoming PR for content debt.

ericvmw replied:
Sounds good, I'll keep it here for now.

tenthirtyam commented on .web-docs/components/builder/vsphere-supervisor/README.md:
```diff
@@ -83,6 +94,39 @@ There are various configuration options available for each step in this builder.
+
+#### Source VM Image Importing
+
+**Optional:**
```
This layout will be in an upcoming PR for content debt.

ericvmw commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,179 @@
+)
+
+type ValidateImportConfig struct {
+	// The remote URL where the to-be-imported image is hosted, the URL must start with `https`.
+	ImportSourceURL string `mapstructure:"import_source_url"`
```
Hm.. that'll be mostly internal, but I'll remove the https check here for now.

ericvmw commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,179 @@
+	// Name of a writable & import-allowed ContentLibrary resource in the namespace where the image will be imported.
+	ImportTargetLocationName string `mapstructure:"import_target_location_name"`
+	// The type of the imported image.
+	// Defaults to `ovf`. Available options include `ovf`.
```
I would rather leave the update to support iso to later when we're ready to support it, even though the change would be small, as I cannot validate it now.

ericvmw commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,179 @@
+	case "ovf":
+		// If it's already "ovf", do nothing.
+	default:
+		return []error{fmt.Errorf("Unsupported ImportTargetImageType: %s", c.ImportTargetImageType)}
```
Same here, better to add ISO support in a separate PR.

ericvmw commented on builder/vsphere/supervisor/step_validate_import.go:
```diff
@@ -0,0 +1,179 @@
+	return nil
+}
+
+func (s *StepValidateImport) isImportOVFFeatureEnabled(ctx context.Context, logger *PackerLogger) error {
```
Yeah, make it isImportFeatureEnabled instead, though the error from the server would still says OVF support.

ericvmw commented:
> Just one more minor nit from `staticcheck` and marking all conversations as closed.
> 
> Looks good, Eric. 🎉

Updated that one too, and some refactoring per Sai's suggestions, please review again when you get a chance, thanks Ryan!

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,367 @@
+	// The SSL certificate of the remote HTTP server that hosts the to-be-imported image.
+	ImportSourceSSLCertificate string `mapstructure:"import_source_ssl_certificate"`
+
+	// Name of a writable & import-allowed ContentLibrary resource in the namespace where the image will be imported.
```
```suggestion
	// Name of a writable an import-allowed contentl ibrary resource in the namespace where the image will be imported.
```

ericvmw replied:
I presume you mean to change `&` to `and`, updated it.

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,367 @@
+
+	// Name of a writable & import-allowed ContentLibrary resource in the namespace where the image will be imported.
+	ImportTargetLocationName string `mapstructure:"import_target_location_name"`
+	// The type of the imported image.
```
```suggestion
	// The type of imported image.
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,367 @@
+	ImportTargetLocationName string `mapstructure:"import_target_location_name"`
+	// The type of the imported image.
+	// Defaults to `ovf`. Available options include `ovf`.
+	ImportTargetImageType string `mapstructure:"import_target_image_type"`
+
+	// The name of the image import request.
```
```suggestion
	ImportTargetImageType string `mapstructure:"import_target_image_type"`
	// The name of the image import request.
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,367 @@
+	// The timeout in seconds to wait for the image to be imported.
+	// Defaults to `600`.
+	WatchImportTimeoutSec int `mapstructure:"watch_import_timeout_sec"`
+
+	// Whether to clean the image imported in this step. If it is set to true, the imported image will be deleted after
+	// source VM is created and becomes ready.
+	// Defaults to false.
+	CleanImportedImage bool `mapstructure:"clean_imported_image"`
```
```suggestion
	// Clean the imported image. If set to `true`, the imported image ia
    // deleted after source virtual machine is created and becomes ready.
	// Defaults to `false`.
	CleanImportedImage bool `mapstructure:"clean_imported_image"`
```

tenthirtyam commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,367 @@
+	})
+
+	if err != nil {
+		logger.Error("error watching the ContentLibraryItemImportRequest object in Supervisor cluster")
```
```suggestion
		logger.Error("error watching the ContentLibraryItemImportRequest object in supervisor cluster")
```

tenthirtyam commented on builder/vsphere/supervisor/step_watch_source.go:
```diff
@@ -92,7 +94,33 @@ func (s *StepWatchSource) Run(ctx context.Context, state multistep.StateBag) mul
+		logger := state.Get("logger").(*PackerLogger)
+
+		if state.Get(StateKeyImportedImageName) == nil {
+			logger.Info(fmt.Sprintf("Skip cleaning imported image since config %s is not set", StateKeyImportedImageName))
```
```suggestion
			logger.Info(fmt.Sprintf("Skipping clean up of imported image since config %s is not set.", StateKeyImportedImageName))
```

tenthirtyam commented on builder/vsphere/supervisor/step_watch_source_test.go:
```diff
@@ -131,6 +133,64 @@ func newFakeVMObj(namespace, name, vmIP string) *vmopv1alpha1.VirtualMachine {
+	}
+	// Check the output lines from the step runs.
+	expectedOutput := []string{
+		fmt.Sprintf("Skip cleaning imported image since config %s is not set", supervisor.StateKeyImportedImageName),
```
```suggestion
		fmt.Sprintf("Skipping clean up of imported image since config %s is not set.", supervisor.StateKeyImportedImageName),
```

tenthirtyam reviewed (changes requested):
Minor items.

ericvmw commented:
> Minor items.

Thanks for the review, addressed the comments, please take a look again.

dilyar85 commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,364 @@
+
+	// Use the ImageName from create source config as the target item name.
+	s.TargetItemName = s.CreateSourceConfig.ImageName
+	s.KeepInputArtifact = s.CreateSourceConfig.KeepInputArtifact
```
Can we use `TTLSecondsAfterFinished` for the cleanup? Maybe worth exposing an optional config with a default value and passing it when creating the import request CR. I wanted to avoid requiring the `CreateSourceConfig` in this step if possible.

ericvmw replied:
I use this `KeepInputArtifact` as a common config, is there a better way to achieve that, or it is not a good practice here? I could use `TTLSecondsAfterFinished`, just feel it would be simpler for users if it can be achieved with a common config for this builder (instead of multiple different configs for different steps). Thoughts?

dilyar85 replied:
Yes, I agree that a common configuration would be simpler for users, but not sure if that's doable or conventional in custom plugins. Did a quick search in the Packer docs but found nothing related to that. @tenthirtyam do you perhaps know?

dilyar85 replied:
Yep it works for me. Thanks again for taking care of this to provide a better experience, Eric!

ericvmw replied:
Though it works functionally in practice, `make generate` doesn't work and it complains duplicate config names, so it looks it is not supported currently (mapstructure-to-hcl2), I may need to use separate configs instead.

dilyar85 commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,364 @@
+	}
+
+	// Use dry run mode to send an image import creation request to API-Server without applying the resource.
+	err := client.NewDryRunClient(s.KubeWatchClient).Create(ctx, importReq)
+	if err != nil && strings.Contains(err.Error(), importOVFFeatureNotEnabledMsg) {
```
```suggestion
	if err := client.NewDryRunClient(s.KubeWatchClient).Create(ctx, importReq); err != nil && strings.Contains(err.Error(), importOVFFeatureNotEnabledMsg) { ... }
```

dilyar85 commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,364 @@
+	return nil
+}
+
+func (s *StepImportImage) isImportSourceValid() error {
+	if strings.HasPrefix(s.ImportImageConfig.ImportSourceURL, "https://") && s.ImportImageConfig.ImportSourceSSLCertificate == "" {
+		return fmt.Errorf("import request source url certificate is empty")
+	}
+
+	return nil
+}
```
Move this to the `Prepare()` function instead so that we can detect this error earlier (before any steps are executed)?

ericvmw replied:
Sounds good.

dilyar85 commented on builder/vsphere/supervisor/step_watch_source.go:
```diff
@@ -92,7 +94,33 @@ func (s *StepWatchSource) Run(ctx context.Context, state multistep.StateBag) mul
 }
 
-func (s *StepWatchSource) Cleanup(state multistep.StateBag) {}
+func (s *StepWatchSource) Cleanup(state multistep.StateBag) {
+	if cleanImage, ok := state.GetOk(StateKeyCleanImportedImage); ok && cleanImage.(bool) {
+		logger := state.Get("logger").(*PackerLogger)
+
+		if state.Get(StateKeyImportedImageName) == nil {
+			logger.Info(fmt.Sprintf("Skipping clean up of imported image since config %s is not set.", StateKeyImportedImageName))
+			return
+		}
+		importImageName := state.Get(StateKeyImportedImageName).(string)
+
+		logger.Info(fmt.Sprintf("Deleting the imported ContentLibraryItem object %s in namespace %s.",
+			importImageName, s.Namespace))
+		ctx := context.Background()
+		importedImage := &imgregv1.ContentLibraryItem{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      importImageName,
+				Namespace: s.Namespace,
+			},
+		}
+		if err := s.KubeWatchClient.Delete(ctx, importedImage); err != nil {
+			logger.Error("error deleting the ContentLibraryItem object %s: %s", importImageName, err)
+		} else {
+			logger.Info(fmt.Sprintf("Successfully deleted the ContentLibraryItem object %s in namespace %s.",
+				importImageName, s.Namespace))
+		}
+	}
+}
```
This can be done in the original `import_image` step, as the `Cleanup` functions are called in an reversed order (i.e. VM will be deleted before the imported image). See https://github.com/hashicorp/packer-plugin-sdk/blob/main/multistep/multistep.go#L52-L59 for more info.

ericvmw replied:
I probably missed how the cleanup works, if the source VM is not created successfully, will this image cleanup still be called (by putting in the import-image step? My original thought was to only clean the image if the source VM can be created and ready, so they can debug the issue by using the image if the source VM creation has issues.

dilyar85 replied:
I wouldn't worry about the source VM status for cleaning up the imported image, since it's false by default. If users explicitly set it to true, they would expect it to get cleaned up after the builder completes.

dilyar85 reviewed (commented):
Thanks for addressing all the previous review comments! Have some additional thoughts to hopefully streamline the user experience of using this plugin with the import step.

ericvmw commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,364 @@
+		}
+	}()
+
+	// Skip importing if the import source URL or target location is not specified.
+	if s.ImportImageConfig.ImportSourceURL == "" || s.ImportImageConfig.ImportTargetLocationName == "" {
+		logger.Info("Skipping image import step. Required configurations for the import are not set.")
+		return multistep.ActionContinue
+	}
```
Hm... sounds good, the latter config is expected if the first one is specified.

ericvmw commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,364 @@
+	ImportImageConfig  *ImportImageConfig
+	CreateSourceConfig *CreateSourceConfig
+
+	SourceURL, SourceSSLCertificate, TargetLocationName, TargetItemName, Namespace string
```
Yeah, those are moved over, now they can be accessed directly.

ericvmw commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,364 @@
+	}
+
+	// Use the ImageName from create source config as the target item name.
+	s.TargetItemName = s.CreateSourceConfig.ImageName
```
`image_name` is required today, I feel it may be better to re-use it instead of introducing another optional config and making the `image_name` optional, since it is the same image that will be used to create the source VM, either imported or pre-created.

dilyar85 replied:
I think the fewer required configs we have, the better experience would be for using this plugin. If the import source URL is provided, there is no need for users to specify the image name as well. I am fine if you want to reuse the `image_name` config here, but we should also update it to be an optional config. We could add a check in `config.Prepare()` to tigger an error if neither of the image name or import source URL is set.

ericvmw replied:
Yeah, the configs exposed to users are not really organized by steps, different steps can declare required configs and do pre-checks to make sure needed configs are provided, it may provide better UX with less configs, esp for configs that connect adjacent steps. Agreed, we better make this config optional if import image is configured.

ericvmw replied:
It turns out configs with the same name cannot be used directly except cross-reference, and this is optional step, so better not move configs here, adding an optional config instead.

ericvmw commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,364 @@
+func (c *ImportImageConfig) Prepare() []error {
+	if c.ImportSourceURL == "" || c.ImportTargetLocationName == "" {
+		return nil
+	}
```
Makes sense, updated it.

dilyar85 commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,378 @@
+	}
+
+	var errs []error
+	if c.ImportSourceURL != "" && c.ImportTargetLocationName == "" {
```
The first condition seems unnecessary as the function returns if it's empty above.

ericvmw replied:
Yep, it can be removed.

dilyar85 commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,378 @@
+			importSuccess := false
+			for _, cond := range itemImportReqObj.Status.Conditions {
+				if cond.Type == imgregv1.ContentLibraryItemImportRequestComplete {
+					importSuccess = cond.Status == corev1.ConditionTrue
```
Add a `break` here to avoid checking other conditions once we find import is successful?

ericvmw replied:
Sounds good.

dilyar85 commented on builder/vsphere/supervisor/step_create_source.go:
```diff
@@ -206,6 +204,12 @@ func (s *StepCreateSource) initStep(state multistep.StateBag) error {
 		kubeClient client.Client
 	)
 
+	if s.Config.ImageName == "" {
+		if s.Config.ImageName, ok = state.Get(StateKeyImportedImageName).(string); !ok {
+			return fmt.Errorf("the image name should be specified in config 'image_name' or generated from image import")
+		}
+	}
```
Can we also check if both `s.Config.ImageName` and `state.Get(StateKeyImportedImageName)` are set, and output a warning that the imported image will be skipped in this case?

ericvmw replied:
Agree, it would be helpful to show a message in the output.

dilyar85 reviewed (approved):
LGTM with some final comments left. @ericvmw Thanks again for your patience in addressing all the reviews!

tenthirtyam reviewed (approved):
LGTM! Thansk for your patience @ericvmw and many thanks to @dilyar85 for your reviews as well.

nywilken reviewed (approved):
Deferring to @tenthirtyam and @dilyar85 on this given their already great review. Apologies for the delay on the merge.

nywilken commented:
I manually resolved the conflicts and pushed. Once green I will merge.

lbajolet-hashicorp commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,379 @@
+			}
+			if importSuccess {
+				// Set VM image ref name if the import is successful.
+				state.Put(StateKeyImportedImageName, strings.ReplaceAll(s.ImportItemResourceName, "clitem-", "vmi-"))
```
Looking at this it's not immediately clear why the `ReplaceAll`, could you add a small comment to explain this?

ericvmw replied:
Actually it doesn't need to replace all, though practically there should only one match of the prefix, I'll update it to replace the first match instead.

lbajolet-hashicorp reviewed (approved):
Just left a few nitpicks mostly, but overall LGTM!

Pre-approving so it's not a blocker later

nywilken commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,379 @@
+	importOVFFeatureNotEnabledMsg = "WCP_VMImageService_ImportOVF feature is not enabled"
+)
+
+var IsWatchingImageImport bool
```
Hmm.. it's a good call out that it is exported, not a big deal, but worth mentioning. My bigger concern around this variable is that it is exported but in the function where it is used we use a mutex to lock writes, which furthers the thought that we should probably need to lock this down further.

ericvmw replied:
Good point, this doesn't need to be global, it is mainly used to help testing, I'll make it part of the step instead.

nywilken commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,379 @@
+}
+
+func (s *StepImportImage) Cleanup(state multistep.StateBag) {
+	if state.Get(StateKeyImageImportRequestCreated) == false {
```
hmm... I believe this will always be false as interface{} does not equal the boolean false. We need to use GetOk here. 
```suggestion
	if v, ok := state.GetOk(StateKeyImageImportRequestCreated); !ok || v == false {
```

lbajolet-hashicorp replied:
> hmm... I believe this will always be false as interface{} does not equal the boolean false. We need to use GetOk here.

I'll even suggest doing `!ok || !v` :) That being said, since `v` is an `interface{}` (unless if casted into a bool), doesn't that have the same problem?

nywilken replied:
Ah yes - you need to type assert. I updated.

ericvmw replied:
Good catch, I'll update as suggested.

ericvmw commented on .web-docs/components/builder/vsphere-supervisor/README.md:
```diff
@@ -22,11 +23,16 @@ Example Packer template:
+  import_source_ssl_certificate = "<SSL certificate of the remote HTTPS server, optional, e.g. '-----BEGIN CERTIFICATE-----xxxxx-----END CERTIFICATE-----'>"
+  import_target_location_name = "<target location / content library for the imported image, optional, e.g. 'cl-6066c61f7931c5ef9'>"
+  import_target_image_type = "<target image type, optional, e.g. 'OVF'>"
   image_name = "<Image name of the source VM, e.g. 'ubuntu-impish-21.10-cloudimg'>"
```
Thanks for the review! They are not mutual exclusive, either both are specified or only `image_name` can be specified, in either case the name from `image_name` will be used as the name of the source image (either imported or existing). On the other hand, either import_source or image_name must be provided for the source image.

ericvmw replied:
Actually I just realized that I added a config `import_target_image_ name` to differentiate from the existing `image_name`, I'll update the doc and examples shortly.

tenthirtyam replied:
Eric, has this been rsolved?

ericvmw commented on builder/vsphere/supervisor/step_import_image.go:
```diff
@@ -0,0 +1,379 @@
+
+var IsWatchingImageImport bool
+
+type ImportImageConfig struct {
```
This is following the code pattern in Supervisor plugin where the same step and config are organized in the same file, I may keep it as it is for consistency within this plugin.

nywilken commented:
Hi @ericvmw thank you for addressing the few pieces of feedback. I will merge this today and release the plugin this week. If @tenthirtyam has no objections.
